plugins/org.python.pydev/src/org/python/pydev/editor/correctionassist/heuristics/AssistFString.java

```diff
@@ -0,0 +1,175 @@
+                }
+                if (c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
+                    i = parsingUtils.eatPar(i, null, c) + 1;
+                    break;
+                }
+                if (Character.isJavaIdentifierPart(c)) {
+                    i++;
+                    while (i < parsingUtils.len()) {
+                        c = parsingUtils.charAt(i);
+                        if (Character.isJavaIdentifierPart(c) || c == '.') {
+                            i++;
+                            continue;
+                        } else if (c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
+                            i = parsingUtils.eatPar(i, null, c) + 1;
+                            c = parsingUtils.charAt(i);
+                            while (Character.isWhitespace(c)) {
+                                i++;
+                                c = parsingUtils.charAt(i);
+                            }
+                            if (c == '.') {
+                                i++;
+                                continue;
+                            }
+                            break; // inner break
+                        }
+                    }
+                    break; // outer break
+                }
+                i++;
+            }
+
+            FastStringBuffer variablesBuf = new FastStringBuffer();
+            variablesBuf.append(doc.get(partitionEndOffset, i - partitionEndOffset));
+            // get only the variable names and remove the last comma
+            variablesBuf.trim().deleteFirst();
+            if (variablesBuf.trim().startsWith('(')) {
+                variablesBuf.deleteFirst();
+                variablesBuf.deleteLast();
+                variablesBuf.trim();
+            }
+            if (variablesBuf.endsWith(',')) {
+                variablesBuf.deleteLast();
+                variablesBuf.trim();
+            }
+
+            // initialize format output string with the properly size allocation (f + string len + variables len)
+            FastStringBuffer strBuf = new FastStringBuffer(1 + partition.getLength() + variablesBuf.length());
+            strBuf.append('f').append(doc.get(partition.getOffset(), partition.getLength()));
+
+            // get variables without literals
+            String variablesWithoutLiterals = PySelection.getLineWithoutCommentsOrLiterals(variablesBuf.toString());
+            // get variable by variable
+            List<String> variables = new ArrayList<String>();
+            parsingUtils = ParsingUtils.create(variablesBuf);
+            int j = 0;
+            while (j < parsingUtils.len()) {
+                char c = parsingUtils.charAt(j);
+                if (Character.isWhitespace(c) || c == ',') {
+                    j++;
+                    continue;
+                }
+                int commaPos;
+                if (c == '{' || c == '[' || c == '(' || c == '"' || c == '\'') {
```

Comment:
i.e.: if we find the `{` here, we should abort the whole process.

Reply:
(as a dict or set here should not be used -- but note that this is only if this is the `top-level` `{` found... if it's found inside a parenthesis then it should work).

Reply:
i.e.: something as: `'%s' % {'a:10'}` should not give any suggestion, but something as `'%s' % ({'a':10},) should do something as: `f'{{'a':10}}'`.